### ISSUE TITLE
orient_outward for non-convex shapes

### ISSUE THREAD

**Issue by jdumas** [libigl/libigl#880]
I was looking at the code for `igl::orient_outward()` and it strikes me as odd to use the centroid to decide the orientation of the patch. I should work only for convex shapes, or at least surfaces whose centroid lies in its interior (also why isn't the centroid simply computed as `V.colwise().mean()`? Am I missing anything?).

My suggestion would be to first orient the patch coherently using `bfs_orient()`, and then use the signed volume of the mesh to decide whether to flip or not. Computing the signed volume of the mesh can be done by summing up the signed volume of the tets made by each triangle and the origin (or maybe the centroid for better stability).

Also the formula for the signed volume of a tet is pretty simple `dot(p2 - p1, cross(p3 - p1, p4 - p1)) / 6.0;`. Is there a reason for `igl::volume()` to compute a lot of square roots? Better stability maybe?

**Comment by alecjacobson** [comment]
`orient_outward` is a very crude heuristic implemented for a comparison to `igl::embree:: reorient_facets_raycast.h` which implements "A Simple Method for Correcting Facet Orientations in Polygon Meshes Based on Ray Casting" [Takayama et al. 2014]. This one calls `bfs_orient` internally

Signed volume for a triangle mesh is already implemented as the second output of `igl::centroid`. So it'd be an easy switch.

Most overloads of `igl::volume` are implemented using the same formula you cite. The only one using `sqrt`s is the intrinsic version (not sure what the rewrite is to avoid them if you're only given edge lengths; the current implementation is probably not the most numerically stable one).

**Comment by jdumas** [comment]
Ok I see, thanks for the clarification. Then I think it would make sense to rewrite `igl::orient_outward()` as I suggest (no need to pull embree to reorient if the input mesh is clean). Does `igl::volume()` compute the signed volume for every tet then? If so then I would simply add the overload to compute the total volume given a surface triangle mesh (returning a double). Regarding the centroid, the fact that it's using this integral formula is to account for potential holes in the volume right? Otherwise you'd simply compute the average of vertices positions?

**Comment by alecjacobson** [comment]
`igl::volume` computes volumes of each tet in tet mesh `V`,`T`

`igl::centroid` computes _the_ centroid and _the_ (signed) volume for a surface mesh `V`, `F`.

The average vertex positions will be incorrect: it will be the centroid of the set of vertex points (i.e., not accounting for surface area of triangles or volume of the implied solid). Applications needing this can just call `V.colwise().mean()` anyway.

**Comment by jdumas** [comment]
Yes I agree =) (just wasn't sure about the difference). Ok then better to create a new file `signed_volume()` for the surface triangle mesh. Will add that to my todolist.

**Comment by alecjacobson** [comment]
sure... it should probably just call `igl::centroid` under the hood to
avoid duplicate code.

**Comment by jdumas** [comment]
Or the other way around? :p

**Comment by alecjacobson** [comment]
I think there's still some confusion. `igl::centroid` does not create a matrix of centroids. It creates the _one_ centroid.

I don't care if centroid calls signed_volume or the other way around, but neither of them should be creating a `T` list, computing individual tet volumes and adding them up.

**Comment by jdumas** [comment]
Yes my bad. I'm always confused reading the types of templated function (`Eigen::PlainObjectBase<Derivedc>`).

**Comment by jdumas** [comment]
Then we'll just make `igl::signed_volume()` an alias for `igl::centroid()` :p
